kernel/src/sys/sched_mlfq.c

```diff
@@ -0,0 +1,328 @@
+			}
+
+			// We don't need to modify the process queue
+			if (p->next == sc->processes2->next) {
+				break;
+			}
+
+			// We insert the next process between the current one and the one
+			// previously scheduled to be switched to.
+			proc_node_t* previous = p;
+			proc_node_t* next_proc = p->next;
+			proc_node_t* moved = sc->processes2->next;
+
+			previous->next = next_proc->next;
+			next_proc->next = moved;
+			sc->processes2->next = next_proc;
+
+			break;
+		}
+
+		p = p->next;
+	} while (p != sc->processes2);
+
+    sc->processes2 = sc->processes2->next;
+
+    return sc->processes2->process;
+}
+
+//next process for circular list 3
+process_t* sched_mlfq_next3(sched_t* sched) { 
+    sched_mlfq_t* sc = (sched_mlfq_t*) sched;
+
+    proc_node_t* p = sc->processes3;
+
+	// Avoid switching to a sleeping process if possible
+	do {
+		if (p->next->process->sleep_ticks > 0) {
+			p->next->process->sleep_ticks--;
+		} else {
+			// We don't need to switch process
+			if (p->next == sc->processes3) {
+				return sc->processes3->process;
+			}
+
+			// We don't need to modify the process queue
+			if (p->next == sc->processes3->next) {
+				break;
+			}
+
+			// We insert the next process between the current one and the one
+			// previously scheduled to be switched to.
+			proc_node_t* previous = p;
+			proc_node_t* next_proc = p->next;
+			proc_node_t* moved = sc->processes3->next;
+
+			previous->next = next_proc->next;
+			next_proc->next = moved;
+			sc->processes3->next = next_proc;
+
+			break;
+		}
+
+		p = p->next;
+	} while (p != sc->processes3);
+
+    sc->processes3 = sc->processes3->next;
+
+    return sc->processes3->process;
+}
+
+void sched_mlfq_exit(sched_t* sched, process_t* process) {
+    sched_mlfq_t* sc = (sched_mlfq_t*) sched;
+    proc_node_t* p = sc->processes1;
+    proc_node_t* p2 = sc->processes2;
+    proc_node_t* p3 = sc->processes3;
+
+    
+    if(flag=1){  //exit for processes in circular list 2
+        while (p2->next->process != process) {
+           p2 = p2->next;
+        }
+
+        proc_node_t* to_remove = p2->next;
+        p2->next = p2->next->next;
+
+        sc->processes1 = p2;
+
+        kfree(to_remove);
+    }
+    else if(flag=2){    //exit for porcesses in circular list 3
+        while (p3->next->process != process) {
+           p3 = p3->next;
+        }
+
+        proc_node_t* to_remove = p3->next;
+        p3->next = p3->next->next;
+
+        sc->processes1 = p3;
+
+        kfree(to_remove);
+    }
+    else{     //exit for processes in circular list 1
+        while (p->next->process != process) {
+           p = p->next;
+        }
+
+        proc_node_t* to_remove = p->next;
+        p->next = p->next->next;
+
+        sc->processes1 = p;
+
+        kfree(to_remove);
+    }
+}
+
+/* Allocates a mlfq with round robin scheduler.
+ */
+sched_t* sched_mlfq() {
+    sched_mlfq_t* sched = kmalloc(sizeof(sched_mlfq_t));
+    if (timer_get_tick() % 5 != 0) 
```

Comment:
Don't do that here, this function will only be called once, in `init_proc`.

You want to have only one `sched_next` function, and have that function check `timer_get_tick()`.

Reply:
Here I have used `timer_get_tick()` as a timer for round robin

Reply:
You misunderstood, `sched_mlfq` is literally called only once in the whole kernel code, by the line you've added in "proc.c". Its purpose is to build the scheduler, its initialization. Adding timer code here doesn't make sense, because that's something you want to do periodically, i.e. in the `sched_next` you've defined. Having changed that, you'll probably find other things to reorganise in your code.